docs/blog/2025-08-mcp-langchain-with-stepflow.md

```diff
@@ -0,0 +1,369 @@
+```
+
+### Connect to LLMs
+Add OpenAI or Anthropic for enhanced analysis:
+```yaml
+- id: deep_analysis
+  component: /builtin/openai
+  input:
+    model: "gpt-4"
+    messages: [...]
+```
+
+### Database Persistence
+Store research in PostgreSQL:
+```yaml
+- id: save_to_db
+  component: /database/insert
+  input:
+    table: "research_reports"
+    data: { $from: { step: create_report } }
+```
+
+## Performance Considerations
+
+The system is designed for efficiency:
+- **Parallel Execution**: Independent steps run concurrently
+- **Lazy Evaluation**: Components execute only when needed
+- **Resource Pooling**: Connections and processes are reused
+- **Streaming Support**: Large outputs can be streamed
+
+In our tests, a complete research workflow executes in under 2 seconds for typical inputs.
+
+## Lessons Learned
+
+Building this research assistant revealed several insights:
+
+1. **Declarative is Powerful**: YAML workflows are easier to understand and modify than imperative code
+2. **Type Safety Matters**: Schema validation catches errors early
+3. **Modularity Wins**: Reusable components accelerate development
+4. **Orchestration Simplifies**: Complex flows become manageable
+
+## Conclusion
+
+The combination of Stepflow, LangChain, and MCP demonstrates how modern orchestration can simplify AI application development. By separating concerns – orchestration, AI processing, and tool integration – we create systems that are both powerful and maintainable.
+
+This research assistant is just the beginning. The same patterns apply to:
+- Data processing pipelines
+- Content generation systems
+- Automated testing frameworks
+- DevOps automation
+- Business process automation
+
+The key insight? **Orchestration transforms complexity into clarity.**
+
+## Production Ready
+
+The MCP filesystem integration provides reliable file operations with proper error handling and bidirectional communication. The system handles MCP's protocol requirements transparently, allowing workflows to use any MCP-compatible tool server.
+
+## Try It Yourself
+
+The complete code is available in the Stepflow repository:
+```bash
+git clone https://github.com/stepflow-ai/stepflow
+cd stepflow/examples/research-assistant
+
+# Install dependencies and run
+cd ../../stepflow-rs && cargo build --release
+cd ../sdks/python && uv add --group dev langchain-core
+cd ../stepflow-rs
+./target/release/stepflow run \
+  --flow=../examples/research-assistant/workflow.yaml \
+  --input=../examples/research-assistant/input_ai_workflows.json \
+  --config=../examples/research-assistant/stepflow-config.yml
+```
+
+## What's Next?
+
+We're excited about the future of AI orchestration:
+- **Visual Workflow Designer**: Drag-and-drop workflow creation
+- **Component Marketplace**: Share and discover components
+- **Cloud Deployment**: Managed Stepflow service
+- **Enhanced Observability**: Real-time monitoring and debugging
+
+Join us in building the future of AI workflow orchestration!
```

Comment:
Maybe in the mean time, point at an interesting example -- perhaps the production AI one?

Reply:
Good call - teasing that for the next post while adding a link to the example.